js/src/vm/PortableBaselineInterpret.cpp

```diff
@@ -4402,13 +4402,18 @@ static PBIResult PortableBaselineInterpret(JSContext* cx_, State& state,
 
     CASE(CheckReturn) {
       Value thisval = POP().asValue();
-      if (frame->returnValue().isObject()) {
-        PUSH(StackVal(frame->returnValue()));
-      } else if (!ret->isUndefined()) {
+      // inlined version of frame->checkReturn(thisval, result) (js/src/vm/Stack.cpp)
+      // except we call PUSH_EXIT_FRAME before any function call to report an error
+      // and if checkReturn would return false then we `goto error`
+      // and otherwise we PUSH(StackVal(result))
+      HandleValue retVal = frame->returnValue();
+      if (retVal.isObject()) {
+        PUSH(StackVal(retVal));
+      } else if (!retVal.isUndefined()) {
         PUSH_EXIT_FRAME();
-        state.value0 = frame->returnValue();
-        ReportValueError(cx, JSMSG_BAD_DERIVED_RETURN, JSDVG_IGNORE_STACK,
-                         state.value0, nullptr);
+        state.value0 = retVal; // is this necessary?
+        ReportValueError(cx, JSMSG_BAD_DERIVED_RETURN, JSDVG_IGNORE_STACK, retVal,
```

Comment:
In general any function in SM that takes a `cx` returns a `bool` that indicates failure (of some sort, to have been flagged on the context before return), so calling this without a check looks suspicious. In general we'd want `if (!...) goto error`, but here it always returns false (its whole purpose is to flag an error) and we thus always goto error. To keep things tidy wrt this implicit branch folding, we have an idiom `MOZ_ALWAYS_FALSE(...); goto error;` that asserts in non-release builds if the thing doesn't actually return false. This is visible for a few other opcodes here as well (e.g. `Throw` should do it).

Reply:
The original `checkReturn` ignores the return value of `ReportValueError` and explicitly does `return false;` afterward. Do you still want me to add `MOZ_ALWAYS_FALSE` around the call in the inlined pbl version?

Reply:
Ah, yeah, the code may not be perfectly consistent with that idiom, but let's keep it consistent at least within PBL.